### ISSUE TITLE
Fix IssueComment handler to handle pull request comments

### ISSUE THREAD

**Issue by joshsmith** [code-corps/code-corps-api#1082]
# Problem

Currently the `Event.IssueComment` handler has an `operational_multi` that matches the following payload:

```
%{"action" => action, "issue" => _, "comment" => _}
```

Unfortunately, this makes the assumption that every `IssueComment` is a comment on a `GithubIssue` and _not_ potentially a comment on a `GithubPullRequest`. This can lead to us creating a `GithubIssue` even though a `GithubPullRequest` already exists.

In the latter case, the payload has an additional `"pull_request":` key in the JSON payload, nested inside the `"issue":` key. We could potentially use this to match against the pull request.

An additional idea: we could also try, as an additional safety measure, to `find_or_create` by looking for _both_ `GithubIssue` and `GithubPullRequest` records by the `github_id` of the issue. This could help us avoid a bad situation because it would effectively enforce uniqueness of `github_id` across both `GithubPullRequest` and `GithubIssue` records.

**Comment by begedin** [comment]
Note that I added a comment to https://github.com/code-corps/code-corps-api/issues/1004#issuecomment-337556757 because it's no longer completely clear to me what our aim is.

However, while the answer to that comment seems to affect the overall process slightly, it does sound like we ought to do the following regardless

- try and link to a `GithubPullRequest`, (find or create, only if issue payload is for a PR)
- failing that, try and link to a `GithubIssue` (find or create)
- if we managed to link to pull request, sync comments with `Task` associated to that pull request
- if we managed to link to an issue, proceed as before

Based on this, we would have to 

- add another operational multi, which matches against a pull request containing payload with the following steps 
  - `:repo`- same as other multi
  - `:github_pull_request` - find or create a github pull request, we can use `PullRequestLinker`
  - `:task_user` - we can use `PullRequest.UserLinker`
  - `:tasks`- we can use `PullRequest.TaskSyncer`
  - `:comment_user` - same as the other multi
  - `:comments` - same as other multi

**Comment by begedin** [comment]
I tried to make a write up off the solution I proposed, and as you said, the problem is the assumption that every comment is for a github issue, not a pull request.

In practice, that means we've modeled our `GithubComment` to always belong to a `GithubIssue`, caused in part by modeling a pull request as a separate entity.

I'm not sure of a good way to deal with it. Simply put, our model diverge from github models in this.

We could add a `belongs_to :github_issue` to the pull request, but we then have to consider how to match it all with tasks

We could allow the comment to belong to a pull request as an alternative to an issue. Require just one of those two relationships, but not both. _That would probably be the simplest we could do, but still not ideal._

We could model it "correctly" by merging the issue and pull request models into one and adding a `type` field. 

I'm having a hard time wrapping my head around this at the moment.

By introducing `GithubComment`, we've actually made it more difficult to handle this. Without it, we could simply

- identify pull request
- get pull request's task
- create comment on that task

Because `github_comment` HAS to belong to an issue, we can no longer do that.

I still think having a `github_comment` is the correct approach, but it may necessarily mean we need to "correct" our other approaches to.

**Comment by begedin** [comment]
Thinking of the "polymorphic comment belongs to" solution further, I think in that case, we absolutely need to differentiate between issues closed by a pull request and issues that are a pull request.

Right now, for a new pull request webhook

- we create a `GithubPullRequest` record
- we create a `Task`  record, belonging to pull request, on each project belonging to repo
- the `Task` records do not belong to a `GithubIssue`

For a new issue webhook

- we create a `GithubIssue`
- we create a `Task`, belonging to issue for each project connected to issue's repo
- the `Tasks` do not belong to pull requests

For a new comment webhook

- we find or match a `GithubIssue`
- we find or match `Task` records, as if an issue webhook
- we find or match a single `GithubComment`, belonging to `GithubIssue`
- we find or match `Comment` records, belonging to `Tasks` and `GithubComment`

For a new pull request comment webhook, with the proposed solution we would

- find or match a `GithubPullRequest`
- find or match a  `GithubComment` belonging to `GithubPullRequest`
- find or match `Tasks` created due to that pull request directly (not due to closing keywords we intend to support)
- we find or match `Comment` records, belonging to `Tasks` and `GithubComment`

That would mean a `Task`

- `belongs to :pull_request`
- `has_many :closing_pull_requests`

On the other end a `GithubPullRequest` 

- `has_many :tasks`
- `has_many :tasks_closed_by`

Meaning we would need a `PullRequestClosingTask` relationship

**Comment by begedin** [comment]
Thinking even further, maybe we're going to far with coupling our models with github models.

Maybe tasks should just not deal with pull requests directly.

- a `GithubIssue` can be a pull request
- a `GithubIssue` can also be referred to by other `GithubIssue`s, which can also be pull requests

```Elixir
schema "github_issues" do
  belongs_to :pull_request, null: true
  has_many :github_issue_referrs_to
  has_many :github_issue_referred_by
end

schema "github_pull_requests" do
  belongs_to :github_issue
end

schema "github_issue_reference" do
  belongs_to :referred_by
  belongs_to :referrs_to
  field :type :string # regular, closing
end

schema "tasks" do
  belongs_to :github_issue
end
```

Then

- a github issue is a pull request if it has a pull request relationship
- a task is rendered as a pull request if associated to an issue which is a pull request
- kind of like waffle, a task card renders pull requests attached if the linked issue has pull request references

Basically, the task is linked to github by issue only, nothing else

**Comment by joshsmith** [comment]
From a conversation today, we've decided that:

- `Task` needs to have _only_ `GithubIssue`, _but_ have `GithubPullRequest` through the issue
- A `GithubIssue` is either 1:0 or 1:1 with a `GithubPullRequest`
- We'll add `github_issue_references` as above
- A `GithubComment` _only_ has a reference to a `GithubIssue` and _not_ a `GithubPullRequest`

**Comment by joshsmith** [comment]
When syncing an issue:
  - without `pull_request`:
    - create or update the issue
  - with `pull_request`:
    - create or update the pull request
    - create or update the issue with the pull request id
